extensions/default/src/MergeDataSource/index.js

```diff
@@ -0,0 +1,195 @@
+import { DicomMetadataStore, IWebApiDataSource } from '@ohif/core';
+import { get } from 'lodash';
+
+/**
+ * Returns deduplicated data by comparing the incomingData with the mergedData based on the mergeKey.
+ * @param {Object} options - The options object.
+ * @param {Array} options.mergedData - The global data array.
+ * @param {Array} options.incomingData - The incoming data array.
+ * @param {string} options.mergeKey - The merge key.
+ * @returns {Array} The deduplicated data array.
+ */
+const getDedupedData = ({ mergedData, incomingData, mergeKey }) => {
+  const keys = mergedData.map(r => get(r, mergeKey));
+  return incomingData.filter(r => !keys.includes(get(r, mergeKey)));
+};
+
+/**
+ * Processes deduplicated data and pushes it to mergedData array.
+ * @param {Object} options - The options object.
+ * @param {string} options.sourceName - The name of the data source.
+ * @param {Array} options.mergedData - The global data array.
+ * @param {Array} options.incomingData - The incoming data array.
+ * @param {string} options.mergeKey - The key to merge the data on.
+ * @param {Function} options.tagFunc - The function to tag the data with.
+ */
+const processDedupedData = ({ sourceName, mergedData, incomingData, mergeKey, tagFunc }) => {
+  let data = [];
+  if (mergeKey) {
+    const dedupedData = getDedupedData({
+      mergedData: mergedData.flat(),
```

Comment:
You should not be calling flat here - the UID keys are always at the top level.  Anything below that is extraneous and might result in missing data if a sub-sequence has an item in it.  Also makes it REALLY slow because in DICOM JSON you will have 1000+ lists in each instance, so you can end up comparing tens of thousands of instances in a linear compare.

Reply:
Can you suggest where to execute flat then? this call is called once per series. I'm not sure why this will be called for every instance.